### ISSUE TITLE
[BUG] Better Intellisense for .NET MAUI XAML files

### ISSUE THREAD

**Issue by danardelean** [microsoft/vscode-dotnettools#565]
### Describe the Issue

Right now the Intellisense for the .NET MAUI XAML is minimal and it cannot be easily used to replace the editing capabilities that Visual Studio for MAC has (cannot find properties if never used, sometimes gives wrong suggestions for the control : for example the xmlns autocomplete for a control in the project is suggested wrong and it won't even compile the project.

We need better support for the .NET MAUI XAML files

### Steps To Reproduce

Both Release and Preview versions have the same problems

### Expected Behavior

I would expect to work as the code editor Visual Studio for MAC has

### Environment Information

macOS

> VS bug [#1894443](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1894443)

**Comment by mgoertz-msft** [comment]
@danardelean Rest assured, we are working on it. We have an entire Experience for this in our product backlog that we are currently working on. It requires extensive changes to the XAML Language Service engine in VS though to be able to reuse it for VS Code, so this will take a while.

**Comment by mgoertz-msft** [comment]
@mauroa Do you know how to sync this issue to VS Experience [#1777602](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1777602)?

**Comment by Marflage** [comment]
I have a related question: How to configure the intellisense to show only those properties for a XAML tag that are applicable in a given context? For example, if I am using a `Button` then after pressing `CTRL`+`Space`, I would like only those properties to be displayed that the `Button` tag supports; not all of the properties. I get confused as to what property would work for a given tag. I guess what I am trying to achieve is the way `.` notation works for an instance variable in any programming language. Does anybody know a way to achieve that? I am new to XAML and .NET MAUI.

**Comment by mgoertz-msft** [comment]
@Marflage That's exactly the purpose of the real XAML Language Service and how it works in VS. We are working on bringing that experience to VS Code. In the time being you are better off using VS or VS Mac.

**Comment by AesBarant** [comment]
I look forward to the innovations you will bring for XAML VsCode

**Comment by arkalyanms** [comment]
> @mauroa Do you know how to sync this issue to VS Experience [#1777602](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1777602)?

I believe Microsoft foundation wont be directly linkable in the azdo bug unfortunately.

**Comment by timheuer** [comment]
@mgoertz-msft Why did we close this? Can we keep this open as the public tracking one?

**Comment by mcyenikoylu** [comment]
Version: 1.85.1 (Universal)
Commit: 0ee08df0cf4527e40edc9aa28f4b5bd38bbff2b2
Date: 2023-12-13T09:48:06.308Z
Electron: 25.9.7
ElectronBuildId: 25551756
Chromium: 114.0.5735.289
Node.js: 18.15.0
V8: 11.4.183.29-electron.0
OS: Darwin x64 22.6.0

MacOS Ventura 13.6.3 (22G436)

As I write code, it memorizes it and brings only what is stored in its memory.

https://www.youtube.com/embed/B0kuOCWXkas?si=eniObTOrSSZdZjI9

**Comment by roughiain** [comment]
Does this include intellisense for bindings?

**Comment by mgoertz-msft** [comment]
@roughiain What we are working on will be powered by the same underlying IntelliSense engine as in the regular Visual Studio IDE, so while the UI in VS Code may be different the actual IntelliSense completions should be the same for both products including Bindings, StaticResources, etc.

And for Binding IntelliSense to work we need to have a `BindingContext` or `x:DataType` set in XAML to provide us the shape of the object being bound to, but that's no different from the regular Visual Studio IDE.

**Comment by ravitejaavv** [comment]
@mgoertz-msft do we have any update on Intellisense auto suggestions for XAML files?

**Comment by Dry-Rusk** [comment]
> @mgoertz-msft do we have any update on Intellisense auto suggestions for XAML files?

+1

**Comment by mgoertz-msft** [comment]
@ravitejaavv, @Dry-Rusk, tl;dr It's still under development.

There are multiple things that need to happen in Roslyn first, one of which just completed yesterday: https://github.com/dotnet/roslyn/pull/71758

This now needs to be inserted back into VS where the classic XAML IntelliSense code lives, so we can start consuming it.

However, there is another issue we are blocked on: https://github.com/dotnet/roslyn/issues/71114 because ever since Roslyn started using its own internal LSP protocol types in December we are no longer able to register LSP handlers for XAML or use External Access APIs with the public VS LSP Protocol types.

**Comment by ravitejaavv** [comment]
> @ravitejaavv, @Dry-Rusk, tl;dr It's still under development.
> 
> There are multiple things that need to happen in Roslyn first, one of which just completed yesterday: [dotnet/roslyn#71758](https://github.com/dotnet/roslyn/pull/71758)
> 
> This now needs to be inserted back into VS where the classic XAML IntelliSense code lives, so we can start consuming it.
> 
> However, there is another issue we are blocked on: [dotnet/roslyn#71114](https://github.com/dotnet/roslyn/issues/71114) because ever since Roslyn started using its own internal LSP protocol types in December we are no longer able to register LSP handlers for XAML or use External Access APIs with the public VS LSP Protocol types.

Thanks for the update. Is there any tentative timeline to close the development ?

**Comment by mgoertz-msft** [comment]
I wouldn't want to speculate on when we would release the first preview. We have run into a few surprises along the way that turned out to be quite costly timewise. But we will certainly make it known when there is something for everyone to try.

**Comment by EntityBox** [comment]
> @Marflage That's exactly the purpose of the real XAML Language Service and how it works in VS. We are working on bringing that experience to VS Code. In the time being you are better off using VS or VS Mac.

VS MAC was retired and does not support .NET 8 - As a MAC user, I'm left with no option for Intellisense. This is a huge setback on coding. Might have been easier to add .Net 8 Support on VS MAC whilst this is being sorted. **Sigh**

Now we code a .Net 7 VS MAC project, copy the code to VS Code for .Net 8 - duplicating efforts and wasting time.

I hope this gets pushed to the priority list soon after the Roslyn fixes.

**Comment by mgoertz-msft** [comment]
FYI, VS Mac is not out of support yet, and it does have a preview option for using it with .NET 8, so you should be able to use it in the meantime.

**Comment by EntityBox** [comment]
@mgoertz-msft - WOW, searched for it and enabled it. **Thanks so much.**

For others looking for it:

![image](https://github.com/microsoft/vscode-dotnettools/assets/120500916/e26ac8eb-fa62-45ef-a936-1e67008c180e)

**Comment by mgoertz-msft** [comment]
The issue may be stale but the work is still progressing regardless. :)

**Comment by mgoertz-msft** [comment]
I'm happy to announce that we have released the first preview of XAML IntelliSense for .NET MAUI today.

![image](https://github.com/microsoft/vscode-dotnettools/assets/14282894/3c73bab1-4dee-4789-b32b-6ce2ff827877)

To try it out please follow these steps:
- Install the current **pre-release** of the C# extension - v2.30.24 (pre-release)
- Install the current release of the C# Dev Kit - v1.5.20
- Install the current **pre-release** of the .NET MAUI extension - v0.11.87 (pre-release)

Let us know what you think. Enjoy!

**Comment by mcyenikoylu** [comment]
@mgoertz-msft I'm so glad to hear this and tested it now. everything works as it should. This is great! :) thanks..

My environment:
- MacOS Sonoma 14.4.1 (23E224)
- VS Code Version: 1.89.1 

Extensions:
- C# v2.28.11
- C# Dev Kit v1.5.20
- .NET MAUI v0.10.61

**Comment by mgoertz-msft** [comment]
And thank you @timheuer for finding the first bug. The editor.onTypeFormatting setting causes errors when typing in XAML:

![image](https://github.com/microsoft/vscode-dotnettools/assets/14282894/7cf419ca-6648-40cc-bb12-931aeaafc5cc)
leads to
![image](https://github.com/microsoft/vscode-dotnettools/assets/14282894/33e613d4-e863-4351-9538-d194f57c5051)

**Comment by mgoertz-msft** [comment]
@mcyenikoylu Your C# extension version looks too old. The first one with XAML IntelliSense support is today's prerelease version:  v2.30.24 (pre-release)

If you're on the release channel for the C# extension try to switch to prerelease. Otherwise, restart VS Code to see if you're offered to update the C# extension.

Of course, that makes me wonder how "everything works as it should"...?

**Comment by singhwong** [comment]
Even though I used the required VS Code extension version, Xaml IntelliSense still didn't work. 

Environment:
MacOS Sonoma 14.4.1 (23E224)
VS Code Version: 1.89.1

**Comment by mgoertz-msft** [comment]
And it's now available in Release, so Switch to Release and grab the latest updates:
- C# - v2.30.28
- C# Dev Kit - v1.6.8
- .NET MAUI - v0.11.96

Enjoy bug fixes, including @timheuer's from above, and rich completion item tooltips:

![image](https://github.com/microsoft/vscode-dotnettools/assets/14282894/a80bea26-bb96-4d5e-b92f-fb23923f326a)

**Comment by mgoertz-msft** [comment]
@mauroa This has been released. Do you have permissions to close this issue?

**Comment by mgoertz-msft** [comment]
@Flyingbee200 No, the XAML IntelliSense for .NET MAUI in VS Code does not work for WPF. It's been requested in issue #1789 **XAML Language Server for WPF**

I'm not aware of an extension that does that.

**Comment by lextm** [comment]
@Flyingbee200 [That kind of extension is now on Marketplace](https://marketplace.visualstudio.com/items?itemName=lextudio.vscode-wpf).
